File: src/1-500/138/CopyListWithRandomPointer.py
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

class Solution:
    def copyRandomList(self, head: 'Node') -> 'Node':
        nodeDict = {}
        dummy = Node(0, None, None)
        nodeDict[head] = dummy
        newHead, pointer = dummy, head
        while pointer:
            node = Node(pointer.val, pointer.next, None)
            nodeDict[pointer] = node
            newHead.next = node
            newHead, pointer = newHead.next, pointer.next
        pointer = head
        while pointer:
            if pointer.random:
                nodeDict[pointer].random = nodeDict[pointer.random]
            pointer = pointer.next
        return dummy.next

solution = Solution()
node1 = Node(1)
node2 = Node(2)
node3 = Node(3)
node4 = Node(4)
node1.next = node2
node2.next = node3
node3.next = node4
node1.random = node4
node4.random = node2
node2.ramdom = node1
node3.random = node3
solution.copyRandomList(node1)

File: src/1-500/138/CopyListwithRandomPointer.java
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/
class Solution {
    //using a hashmap to contruct old node to new node mapping
    public Node copyRandomList(Node head) {
        if (head == null) {
            return null;
        }
        Map<Node, Node> map = new HashMap<>();
        Node helperHead = head;
        while(helperHead != null) {
            Node copiedNode = map.getOrDefault(helperHead, new Node(helperHead.val));
            map.put(helperHead, copiedNode);
            if (helperHead.next != null) {
                Node copiedNext = map.getOrDefault(helperHead.next, new Node(helperHead.next.val));
                copiedNode.next = copiedNext;
                map.put(helperHead.next, copiedNext);
            }
            if (helperHead.random != null) {
                Node copiedRandom = map.getOrDefault(helperHead.random, new Node(helperHead.random.val));
                copiedNode.random = copiedRandom;
                map.put(helperHead.random, copiedRandom);
            }

            helperHead = helperHead.next;
        }
        return map.get(head);
    }
}